manual/src/refman/patterns.etex

```diff
@@ -135,13 +204,26 @@ is a shorthand for the following or-pattern:
 @"(" "`"tag-name_1"(_" ":" typexpr_1")" "|" \ldots "|" "`"tag-name_n"(_"
 ":" typexpr_n"))"@. It matches all values of type @"[<" typeconstr "]"@.
 
+\begin{caml_example}{toplevel}
+type 'a any_rectangle = [`Square of 'a | `Rectangle of 'a * 'a]
+
+let to_string = function
+  | #any_rectangle -> "Rectangle"
+  | `Circle _ -> "Circle";;
```

Comment:
This example is fine, but #-patterns are typically coupled with `as` to refine the type of the input, and I think it would be nice to examplify it here:

```ocaml
type 'a rectangle = [ `Square of 'a | `Rectangle of 'a * 'a ]
type 'a shape = [ `Circle of 'a | 'a rectangle ]

let try_rectangle : 'a shape -> 'a rectangle option = function
  | #rectangle as r -> Some r
  | `Circle _ -> None
```

Reply:
Is the type annotation required in this example to get the behaviour you desire? If so, i'd rather not use it. If not, I'm happy to.

Reply:
It's not required, but it helps figuring out what happens and what we are trying to do. Maybe the toplevel signature output also does that, and then it is okay to drop the annotations. (In generally when I write code using polymorphic variants, I *do* annotate carefully, because in my experience it catches my mistakes faster and more locally.)